spec/todoApp.spec.js

```diff
@@ -0,0 +1,177 @@
+    const result = todoApp.showAll();
+    //verify
+    expect(result).toEqual(errorMessage);
+  });
+
+  it("method setComplete() sets todo item's status to complete", () => {
+    // setup
+    const task1 = "write tests";
+    const todo1 = todoApp.create(task1);
+    // execute
+    const result = todoApp.setComplete(todo1.id);
+    //verify
+    expect(result.status).toEqual("complete");
+  });
+
+  it("shows list of incomplete todos if there are any in array", () => {
+    // setup
+    const task1 = "write tests";
+    const task2 = "write better tests";
+    const task3 = "write even better tests";
+    const todo1 = todoApp.create(task1);
+    const todo2 = todoApp.create(task2);
+    const todo3 = todoApp.create(task3);
+    todoApp.setComplete(todo2.id);
+    const incompleteTodoList = [todo1, todo3];
+
+    // execute
+    const result = todoApp.showIncomplete();
+
+    //verify
+    expect(result).toEqual(incompleteTodoList);
+  });
+
+  it("shows list of complete todos if there are any in array", () => {
+    // setup
+    const task1 = "write tests";
+    const task2 = "write better tests";
+    const task3 = "write even better tests";
+    const task4 = "write automated tests";
+    const todo1 = todoApp.create(task1);
+    const todo2 = todoApp.create(task2);
+    const todo3 = todoApp.create(task3);
+    const todo4 = todoApp.create(task4);
+    todoApp.setComplete(todo1.id);
+    todoApp.setComplete(todo2.id);
+    const completeTodoList = [todo1, todo2];
+
+    // execute
+    const result = todoApp.showComplete();
+
+    //verify
+    expect(result).toEqual(completeTodoList);
+  });
+
+  it("returns todo item if it exists in todo list", () => {
+    // setup
+    const task = "get some water!";
+    const todo = todoApp.create(task);
+    // execute
+    const result = todoApp.searchById(1);
+    // verify
+    expect(result).toEqual(todo);
+  });
+
+  it("returns message if todo item doesn't exist in todo list", () => {
+    // setup
+    const message = "The task doesn't exist!";
+    // execute
+    const result = todoApp.searchById(1);
+    // verify
+    expect(result).toEqual(message);
+  });
+
+  it("returns a message when todo item is removed", () => {
+    // setup
+    const task1 = "get some water!";
+    const task2 = "get a little break!";
+    const todo1 = todoApp.create(task1);
+    const todo2 = todoApp.create(task2);
+    // execute
+    const result = todoApp.removeById(todo2.id);
+    // verify
+    expect(result).toEqual("Todo with id:2 was removed successfully!");
+  });
+
+  it("returns a full text description if there is a single todo item", () => {
+    // setup
+    const task = "write some good code";
+    const todo = todoApp.create(task);
+    const expected = [todo];
+    // execute
+    const result = todoApp.showAll();
+    // verify
+    expect(result).toEqual(expected);
+  });
+
+  it("shows first 20chars of a todo item text if there are more than one item in the list", () => {
```

Comment:
How can you improve the quality of this assertion? At the moment the character length of the items is 20chars or less - but we're asserting we only show up to 20 when showing all.

> make sure you have text that is more than 20chars and then assert that it's reduced.

Reply:
In this case the assertion could be `reduces text up to first 20chars if the text length is greater than 20`. So then the algorithm for the test would be: 
1. check if the length of the text to create is greater than 20, using `toBeGreaterThan()` matcher
2. call `todoApp.create()` method to create the todo item
3. call `todoApp.showAll()` method, that we're testing
4. verify the result to be equal to 20
Then probably the other step would be to create another test with assertion `returns full text if it's length is less than 20chars`, though not sure if it's needed?